File: examples/assertion-failure/symex.c
#include <stdint.h>
#include <stddef.h>

static volatile uint32_t* const SYMCTRL_ADDR = (uint32_t*)0x02020000;
static volatile uint32_t* const SYMCTRL_SIZE = (uint32_t*)0x02020004;
static volatile uint32_t* const SYMCTRL_CTRL = (uint32_t*)0x02020008;

#define SYMEX_ERROR (1 << 31);
#define SYMEX_EXIT  (1 << 30);

void
make_symbolic(void *ptr, size_t size)
{
	*SYMCTRL_ADDR = (uintptr_t)ptr;
	*SYMCTRL_SIZE = size;
}

void
symex_error(void)
{
	*SYMCTRL_CTRL = SYMEX_ERROR;
}

void
symex_exit(void)
{
	*SYMCTRL_CTRL = SYMEX_EXIT;
}

File: examples/zig-out-of-bounds/symex.zig
const SYMEX_CTRL_ADDR: usize = 0x02020000;
const SYMEX_CTRL_SIZE: usize = 0x02020004;
const SYMEX_CTRL_CTRL: usize = 0x02020008;

const SYMEX_ERROR: u32 = 1 << 31;
const SYMEX_EXIT:  u32 = 1 << 30;

pub const SymEx = struct {
    pub fn exit(self: SymEx) void {
        const ptr = @intToPtr(*volatile u32, SYMEX_CTRL_CTRL);
        ptr.* = SYMEX_EXIT;
    }

    pub fn panic(self: SymEx) void {
        const ptr = @intToPtr(*volatile u32, SYMEX_CTRL_CTRL);
        ptr.* = SYMEX_ERROR;
    }

    pub fn make_symbolic(self: SymEx, addr: usize, size: u32) void {
        const aptr = @intToPtr(*volatile u32, SYMEX_CTRL_ADDR);
        const sptr = @intToPtr(*volatile u32, SYMEX_CTRL_SIZE);

        aptr.* = addr;
        sptr.* = size;
    }
};
